src/chainrules.jl

```diff
@@ -1,121 +1,152 @@
 # ffts
-function ChainRulesCore.frule((_, Δx, _), ::typeof(fft), x::AbstractArray, dims)
-    y = fft(x, dims)
-    Δy = fft(Δx, dims)
+# we explicitly handle both unprovided and provided dims arguments in all rules, which
+# results in some additional complexity here but means no assumptions are made on what
+# signatures downstream implementations support.
+function ChainRulesCore.frule(Δargs, ::typeof(fft), x::AbstractArray, dims=nothing)
```

Comment:
I'm not happy about this PR because it means the signature of the AD rules is different from the signatures of `fft` etc. - we do not support `dims = nothing` in any of these methods.

Reply:
A default keyword argument simply expands to separate dispatches on the signatures `fft(x, dims)` and ``fft(x)`. The `dims=nothing` is just a way of sharing logic in these cases

Reply:
So I would not say the signatures are different?

Reply:
My point is: You can call `frule(.., fft, x, nothing)` but you cannot call `fft(x, nothing)`. This breaks the correspondence between the primal function and the rules, and makes the signatures inconsistent.

There is no clean way to share code as long as `fft(x)` and `fft(x, dims)` are completely separate. Introducing `fft(x) = fft(x, 1:ndims(x))` or `fft(x) = fft(x, nothing)`, and demanding that downstream packages implement `fft(x, dims)` only would solve these issues. Otherwise you have to copy the code or use something like `@eval` to do it for you.

Reply:
That's a fair point, I didn't realize the `nothing` case. Sharing code would be easy enough with a shared helper function, e.g. replacing my current function with something like `_fft_rrule` and calling it in both cases, so that all the dispatches are correct. If you're opposed to that,  I can look into how to modify `src/definitions.jl` to support your solution.